Add yamllint checking

obnoxxx opened the pull request:
This PR adds yaml lint checking to the tests and to the github  CI workflows.
The code is is inspired by but not identical to  the code from samba-operator.

This PR is proposed to allow for easier update of our yamls, e. g. workflow yamls in future PRs.

obnoxxx commented:
gitlint failed because gitlint was not found. trying to fix it now ...

synarete reviewed (commented):
The two commits looks perfectly fine, but as a follow-up commit there should be a fix to all those yaml files (probably using `hack/yq-fixup-yamls.sh`) or else `make check` will fail.

phlogistonjohn commented:
According to the results I looked at the tool *is* found and does not like the length of the line:
```
essentially taken from corresponding code in https://github.com/samba-in-kubernetes/samba-operator
```
I think if you used `git rebase -i` and reworded that commit to add a line break right before the URL your commit would pass the check

phlogistonjohn commented:
Changes generally look OK to me, although I think we'll need longer term cleanups to the makefile (not for this PR). I will approve once the commit message is fixed

obnoxxx commented:
@phlogistonjohn  wrote:
> According to the results I looked at the tool _is_ found and does not like the length of the line:
> 
> ```
> essentially taken from corresponding code in https://github.com/samba-in-kubernetes/samba-operator
> ```
> 
> I think if you used `git rebase -i` and reworded that commit to add a line break right before the URL your commit would pass the check


Thanks for pointing this out! I fixed it but my attempts to fix the installation of gitlint do not seem to have been successful: now th PR's ci check run shows the same error I see locallyt:

```
0s
Run make check-gitlint
env: ‘GOBIN’: No such file or directory
/home/runner/work/samba-container/samba-container/hack/install-tools.sh --gitlint /home/runner/work/samba-container/samba-container/.bin
Error: [Errno [1](https://github.com/samba-in-kubernetes/samba-container/actions/runs/5044843692/jobs/9048383923?pr=132#step:4:1)3] Permission denied: '/bin/.py'
make: *** [Makefile:3[6](https://github.com/samba-in-kubernetes/samba-container/actions/runs/5044843692/jobs/9048383923?pr=132#step:4:7)2: /home/runner/work/samba-container/samba-container/.bin/gitlint] Error 1
Error: Process completed with exit code 2
```

obnoxxx commented:
@phlogistonjohn  wrote:
> Changes generally look OK to me, although I think we'll need longer term cleanups to the makefile (not for this PR). I will approve once the commit message is fixed

I think the commit message is fixed now, but my attempt to fix the mechanism to install gitlint has failed ...

phlogistonjohn commented:
You can't just copy the install-tools.sh from the other repo. First, off a lot of the tools in that script are not relevant to this repo, second I was trying to fix issues with the other script here first as a test bed so just bringing everything over wholesale undoes what I was trying to accomplish.

obnoxxx commented:
> > @phlogistonjohn wrote:
> > > According to the results I looked at the tool _is_ found and does not like the length of the line:
> > > ```
> > > essentially taken from corresponding code in https://github.com/samba-in-kubernetes/samba-operator
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I think if you used `git rebase -i` and reworded that commit to add a line break right before the URL your commit would pass the check
> > 
> > 
> > Thanks for pointing this out! I fixed it but my attempts to fix the installation of gitlint do not seem to have been successful: now th PR's ci check run shows the same error I see locallyt:
> > ```
> > 0s
> > Run make check-gitlint
> > env: ‘GOBIN’: No such file or directory
> > /home/runner/work/samba-container/samba-container/hack/install-tools.sh --gitlint /home/runner/work/samba-container/samba-container/.bin
> > Error: [Errno [1](https://github.com/samba-in-kubernetes/samba-container/actions/runs/5044843692/jobs/9048383923?pr=132#step:4:1)3] Permission denied: '/bin/.py'
> > make: *** [Makefile:3[6](https://github.com/samba-in-kubernetes/samba-container/actions/runs/5044843692/jobs/9048383923?pr=132#step:4:7)2: /home/runner/work/samba-container/samba-container/.bin/gitlint] Error 1
> > Error: Process completed with exit code 2
> > ```
> 
@synarete  wrote:
> @obnoxxx There are few issues with opt-level Makefile, in particular when is uses and pass env and args to `hack/install-tools.sh` helper script. Please take a look at my fixup proposal: [synarete@462e571](https://github.com/synarete/samba-container/commit/462e57182f9ea16438e27f884dd1575db5aa630a)

Thanks! I took that commit and squashed it with mine. Now it is failing differently. looking into it now ...

obnoxxx commented:
updated the PR so that things work locally for me. now there is a gitlint invocation error here in ci ... @phlogistonjohn , @synarete

phlogistonjohn commented:
The problem is the new line:
```
check: check-shell-scripts check-yaml check-gitlint
```

Remove check-gitlint from this line and it will stop failing.  The issue is that github doesn't normally pull the the various branches and tags in a repo when using the standard github action for checkout. The `check-commits` action uses additional steps to ensure all the needed branches are present. Therefore the `check-gitlint` rule must not be run by default by `check`

obnoxxx commented:
@phlogistonjohn  wrote:
> The problem is the new line:
> 
> ```
> check: check-shell-scripts check-yaml check-gitlint
> ```
> 
> Remove check-gitlint from this line and it will stop failing. The issue is that github doesn't normally pull the the various branches and tags in a repo when using the standard github action for checkout. The `check-commits` action uses additional steps to ensure all the needed branches are present. Therefore the `check-gitlint` rule must not be run by default by `check`


Thanks for the analysis! I undid that line change and now it stopped failing indeed.

FWIW, I did not really mean to suggest to merge the PR with all included commits. instead, I offer to squash  a few commits, specifically  these:
* https://github.com/obnoxxx/samba-container/commit/cc479df1ec6a500273a96258fa3a5efde34b6d1d 
* https://github.com/obnoxxx/samba-container/commit/b71692360b8c795e380a7c3b5ca1f3a2f4a6e127 
* https://github.com/obnoxxx/samba-container/commit/e76a981cbd2d99010ac794bcff3e1b250332cae9

phlogistonjohn commented on hack/install-tools.sh:
```diff
@@ -52,7 +52,7 @@ if [ -z "$PY_CMD" ]; then
 fi
 
-ALT_BIN="$(realpath "${2:-.bin}")"
+ALT_BIN="${2:-.bin}"
```
I'm not thrilled with this because it worked fine on Linux. It's only on MacOS (and maybe bsd?) that this failed. It appears that realpath on linux doesn't require the path to exist, but Mac does.

At the very least could you mention that this is a fix for *Mac* in the commit message?

obnoxxx replied:
done, thanks

phlogistonjohn commented on .yamllint.yaml:
```diff
@@ -0,0 +1,36 @@
+---
```
*Commit message*
```
checks: dd yamllint config file 
```
"dd" should probably be "add"?

obnoxxx replied:
fixed, thanks

phlogistonjohn commented on Makefile:
```diff
@@ -294,6 +301,15 @@ $(DIR)/.common: $(COMMON_DIR)
+# Helper function to re-format yamls using helper script
+define yamls_reformat
+	YQ=$(YQ) $(CURDIR)/hack/yq-fixup-yamls.sh $(1)
+endef
```
I don't see this used anywhere. Please remove it or use it somehow.

phlogistonjohn replied:
Wait a second.... now that I've read the whole PR, is there even a yq-fixup-yamls.sh script to call?
Let's keep this PR lean and drop everything related to YQ for now. Just do the yamllint bits and then in a later PR you can add YQ related stuff.

obnoxxx replied:
right, in this version of the PE, I forgot to copy the script. ... I'll dropt the reformatting for now and we can discuss whether to add it later

obnoxxx replied:
gone now

phlogistonjohn commented on Makefile:
```diff
@@ -294,6 +301,15 @@ $(DIR)/.common: $(COMMON_DIR)
+	YQ=$(YQ) $(CURDIR)/hack/yq-fixup-yamls.sh $(1)
+endef
+
+
+
+
```
Excess whitespace.

obnoxxx replied:
gone now

phlogistonjohn commented on Makefile:
```diff
@@ -253,6 +256,14 @@ test-nightly-server: build-nightly-server
+
+# Format yaml files for yamllint standard
+yaml-fmt: $(filter  $(ALT_BIN)%,(YQ_CMD))
+	$(call yamls_reformat, $(CURDIR))
```
Oh here it is. Honestly, I don't see the point of using call here. This is the only place YQ gets used so there's no need for code reuse.  I'd drop the define from the previous patch and just copy the YQ command to this line.

phlogistonjohn reviewed (changes requested):
Got a few minor things that need cleaning up, nothing major.

obnoxxx commented:
updated once more to fix gitlint errors

phlogistonjohn commented on Makefile:
```diff
@@ -293,7 +302,6 @@ _img_build: $(DIR)/.common
 $(DIR)/.common: $(COMMON_DIR)
 	$(RM) -r $(DIR)/.common
 	cp -r $(COMMON_DIR) $(DIR)/.common
-
```
You've deleted an existing blank between targets, I think this will hurt readability.

obnoxxx replied:
fixed, thanks!

phlogistonjohn commented on Makefile:
```diff
@@ -248,14 +248,23 @@ test-nightly-server: build-nightly-server
+check: check-shell-scripts check-yaml
 .PHONY: check
 
+
+
+
+
```
Excess whitespace.

obnoxxx replied:
fixed, thanks

phlogistonjohn reviewed (changes requested):
Still got some various white space issues.

phlogistonjohn reviewed (dismissed):
LGTM, thanks!

obnoxxx commented:
@mergifyio refresh

anoopcs9 commented on Makefile:
```diff
@@ -248,14 +248,18 @@ test-nightly-server: build-nightly-server
 .PHONY: check-shell-scripts
 
+
+check-yaml: $(filter $(ALT_BIN)%,$(YAMLLINT_CMD))
```
Are we expecting `make check-yaml` to fail when _**yamllint**_ command is not available on the system? Because right now _hack/install-tool.sh_ is unaware of `--yamllint` option and fails.

In GitHub CI _**yamllint**_ is present by default as _/opt/pipx_bin/yamllint_.

phlogistonjohn replied:
Hrm. I wonder if we have another case of a disappearing commit because I swear I remember there being a commit to add `--yamllint` at one point. But good catch regardless!

obnoxxx replied:
This is very confusing, indeed, I am also sure it was there before, after all, the checks all passed in the PR previously... unsure what to do about it... 🤷🏼

anoopcs9 replied:
> In GitHub CI _**yamllint**_ is present by default as _/opt/pipx_bin/yamllint_.

and that's why it passes on GitHub.

We'll have to resurrect the change within _hack/install-tools.sh_ script to install **_yamllint_** if not present on the system.

obnoxxx replied:
@anoopcs9   even if it was not needed for the CI passing, I re-added the `--yamllint` switch to the install-tool script

obnoxxx replied:
@anoopcs9   even if it was not needed for the CI passing, I re-added the `--yamllint` switch to the install-tool script

obnoxxx replied:
@phlogistonjohn  - FYI

anoopcs9 commented on Makefile:
```diff
@@ -248,14 +248,18 @@ test-nightly-server: build-nightly-server
 
+
+check-yaml: $(filter $(ALT_BIN)%,$(YAMLLINT_CMD))
+	$(YAMLLINT_CMD) -c $(CURDIR)/.yamllint.yaml $(CURDIR) 
```
```suggestion
	$(YAMLLINT_CMD) -c $(CURDIR)/.yamllint.yaml $(CURDIR)
```

trailing whitespace to be removed.

obnoxxx commented:
@mergifyio rebase

anoopcs9 reviewed (approved):
lgtm, thanks.

obnoxxx commented:
@anoopcs9 - thanks for approval! and please ignore my renewed review request: it was by mistake. ... :-)

synarete commented on hack/install-tools.sh:
```diff
@@ -52,16 +61,23 @@ if [ -z "$PY_CMD" ]; then
 fi
 
-ALT_BIN="$(realpath "${2:-.bin}")"
+ALT_BIN="${2:-.bin}"
```
I just realizes that `.bin` is not listed in `.gitignore` at the root of this repo. Should add it for the case where `yamllint` is not installed on local host machine and installed locally within this repo `.bin` directory.

obnoxxx replied:
done. added a commit to do this.

synarete commented on Makefile:
```diff
@@ -273,14 +273,18 @@ test-nightly-server: build-nightly-server
 .PHONY: check-shell-scripts
 
+
+check-yaml: $(filter $(ALT_BIN)%,$(YAMLLINT_CMD))
```
This target failed on my machine due to projects yaml files which are not aligned with required standard. Fixed by using the following command: ```find ./ -type f -name '*.yml' -exec yq eval --inplace {} \;```

Consider fixing as part of this PR.

anoopcs9 replied:
Does it fail for you? Or just warnings? In my case `make check-yaml` ran successfully with warnings. Nevertheless it makes sense to fix all warnings in this PR itself.

obnoxxx replied:
@phlogistonjohn asked me some time ago to ignore the warnings since he is convinced our yamls are all perfectly fine :-)

obnoxxx replied:
@synarete wrote:
> This target failed on my machine due to projects yaml files which are not aligned with required standard.

for me and in the ci, it succeeds with a few warnings.


Fixed by using the following command: `find ./ -type f -name '*.yml' -exec yq eval --inplace {} \;`
> 
> Consider fixing as part of this PR.

I am reluctant to add fixes for things that do not appear to bw broken to me ;-)

phlogistonjohn replied:
Yes, I find yamllint (at least the config in the samba-operator repo) overly picky. I especially have trouble remembering what style of indent it wants for lists. I suggested that @obnoxxx just make a few of the things it finds warnings so we we are alerted to what may not be optimal, but we can ignore it for now.

Another issue is that some of these tools seem to have conflicting ideas of what the optimal yaml formatting is and I don't enjoy having to run one tool to generate yamls and then another to reformat them. That's more of an issue in the operator repo than here, since this repo has no yaml generating tools.

synarete reviewed (commented):
Looks fine. I would add a fix to yml files on-top of this one.,

anoopcs9 commented on hack/install-tools.sh:
```diff
@@ -52,16 +61,23 @@ if [ -z "$PY_CMD" ]; then
+ALT_BIN="${2:-.bin}"
 case "$1" in
-    --gitlint)
+    "--yamllint")
```
May be you could add `--yamllint` to the help section listed towards the end of script.
```diff --git a/hack/install-tools.sh b/hack/install-tools.sh
index 61aec4d..2c7bc48 100755
--- a/hack/install-tools.sh
+++ b/hack/install-tools.sh
@@ -74,5 +74,6 @@ case "$1" in
         echo "available tools:"
         echo "  --gitlint"
         echo "  --shellcheck"
+        echo "  --yamllint"
     ;;
 esac
```

obnoxxx replied:
good catch! done, thanks!

synarete reviewed (dismissed):
lgtm

anoopcs9 commented on Makefile:
```diff
@@ -273,24 +273,29 @@ test-nightly-server: build-nightly-server
 
-clean:
+clean: clean-buildfiles clean-altbin
+clean-buildfiles:
 	$(RM) $(BUILDFILE_PREFIX)*
-.PHONY: clean
+.PHONY: clean clean-buildfiles
```
Can we bring all clean targets together at one place?

obnoxxx replied:
@anoopcs9  wrote:

> good idea, I will update 



> Can we bring all clean targets together at one place?
>

good idea, I will update

obnoxxx replied:
added a corresponding  commit

anoopcs9 commented on hack/install-tools.sh:
```diff
@@ -73,6 +89,7 @@ case "$1" in
         echo ""
         echo "available tools:"
         echo "  --gitlint"
+	echo " --yamllint" 
```
trailing whitespace
```suggestion
	echo "  --yamllint"
```

anoopcs9 commented on Makefile:
```diff
@@ -273,24 +273,33 @@ test-nightly-server: build-nightly-server
+clean-buildfiles:
+	$(RM) $(BUILDFILE_PREFIX)*
+.PHONY:  clean-buildfiles
+clean-altbin:
+	$(RM) -r $(ALT_BIN)
+.PHONY: clean-altbin
```
Previous occurrence of `clean-altbin` target still exists towards the end of Makefile. I think its better to have everything(clean related rules) towards the end of Makefile.

obnoxxx replied:
fixed, thanks!

synarete reviewed (approved):
lgtm

phlogistonjohn reviewed (approved):
LGTM
